scipy/special/xsf/cdflib.h

```diff
@@ -0,0 +1,79 @@
+
+#pragma once
+
+#include "cephes/igam.h"
+#include "config.h"
+#include "error.h"
+#include "tools.h"
+
+namespace xsf {
+
+XSF_HOST_DEVICE inline double gdtrib(double a, double p, double x) {
+    if (std::isnan(p) || std::isnan(a) || std::isnan(x)) {
+        return std::numeric_limits<double>::quiet_NaN();
+    }
+    if (!((0 <= p) && (p <= 1))) {
+        set_error("gdtrib", SF_ERROR_DOMAIN, "Input parameter p is out of range");
+        return std::numeric_limits<double>::quiet_NaN();
+    }
+    if (!(a > 0) || std::isinf(a)) {
+        set_error("gdtrib", SF_ERROR_DOMAIN, "Input parameter a is out of range");
+        return std::numeric_limits<double>::quiet_NaN();
+    }
+    if (!(x >= 0) || std::isinf(x)) {
+        set_error("gdtrib", SF_ERROR_DOMAIN, "Input parameter x is out of range");
+        return std::numeric_limits<double>::quiet_NaN();
+    }
+    if (x == 0.0) {
+        if (p == 0.0) {
+            set_error("gdtrib", SF_ERROR_DOMAIN, "Indeterminate result for (x, p) == (0, 0).");
+            return std::numeric_limits<double>::quiet_NaN();
+        }
+        /* gdtrib(a, p, x) tends to 0 as x -> 0 when p > 0 */
+        return 0.0;
+    }
+    if (p == 0.0) {
+        /* gdtrib(a, p, x) tends to infinity as p -> 0 from the right when x > 0. */
+        set_error("gdtrib", SF_ERROR_SINGULAR, NULL);
+        return std::numeric_limits<double>::infinity();
+    }
+    if (p == 1.0) {
+        /* gdtrib(a, p, x) tends to 0 as p -> 1.0 from the left when x > 0. */
+        return 0.0;
+    }
+    double q = 1.0 - p;
+    auto func = [a, p, q, x](double b) {
+	if (p <= q) {
+	    return cephes::igam(b, a * x) - p;
+	}
+	return q - cephes::igamc(b, a * x);
+    };
+    double lower_bound = std::numeric_limits<double>::min();
+    double upper_bound = std::numeric_limits<double>::max();
+    auto [xl, xr, f_xl, f_xr, bracket_status] = detail::bracket_root_for_cdf_inversion(
+        func, 1.0, lower_bound, upper_bound, -0.875, 7.0, 0.125, 8, false, 342
```

Comment:
Where do these values come from?

Reply:
Good question; this could use a comment. I wanted the brackets to have the property that (at least approximately) each candidate bracket contains the same number of unique floating point numbers. This makes it so regardless of which bracket is chosen, the worst case situation for the number of function evaluations Chandrupatla needs will remain (at least approximately) constant.

The meaning of the constants is

```C++
1.0 // initial guess
-0.875 // initial step if bracket finder will search to the left
7.0 // initial step if bracket finder will search to the right
0.125 // scale factor when searching to the left
8.0 // scale factor when searching to the right
false // whether function is increasing. can save function evals by knowing this in advance
342 // max iterations. Chosen as small as possible to search all normalized floating point values exhaustively.
```

I worked out this scheme together with @fancidev. Some key points:
* Keeping the scale factor and bracket endpoints as powers of 2 prevents round off error from leading to numerical catastrophe and causing unpredictable behavior.
* A scale factor of 8 (or 1/8 when searching left) offers a good tradeoff between the size of the brackets offered to the root finder, and the number of iterations needed to search all normalized floating point values exhaustively. This was chosen based on examining runtimes for various cases compared to the original cdflib implementation.
* 342 is just $\max\left(\lceil \log_8\left(\mathrm{DBL\textunderscore MAX}\right)\rceil, \lceil \log_{0.125}\left(\mathrm{DBL\textunderscore MIN}\right) \rceil\right)$.

Reply:
That would be a great comment in the code if you're willing.

Yeah, it's good that the factor is bigger than 2. During the original `_bracket_root` PR, I had a theoretical argument for why bigger is better, but we never got around to talking about it.